src/vulkan-renderer/wrapper/query_pool.cpp

```diff
@@ -0,0 +1,176 @@
+        case VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_PRIMITIVES_BIT:
+            if (!static_cast<bool>(m_device_features.geometryShader)) {
+                spdlog::warn("Can't add VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_PRIMITIVES_BIT to pipeline "
+                             "statistics flag bit!");
+                spdlog::warn("Geometry shaders are not available on this gpu (device_features.geometryShader = false)");
+                break;
+            } else {
+                ret_val.emplace_back(VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_PRIMITIVES_BIT);
+            }
+            break;
+        default:
+            // No special check required for this flag.
+            ret_val.emplace_back(flag_bit);
+            break;
+        }
+    }
+
+    ret_val.shrink_to_fit();
+    return ret_val;
+}
+
+std::string QueryPool::get_pipeline_stats_flag_bit_name(const VkQueryPipelineStatisticFlagBits bit) {
+    switch (bit) {
+    case VK_QUERY_PIPELINE_STATISTIC_INPUT_ASSEMBLY_VERTICES_BIT:
+        return "Input assembly vertex count";
+    case VK_QUERY_PIPELINE_STATISTIC_INPUT_ASSEMBLY_PRIMITIVES_BIT:
+        return "Input assembly primitives count";
+    case VK_QUERY_PIPELINE_STATISTIC_VERTEX_SHADER_INVOCATIONS_BIT:
+        return "Vertex shader invocations";
+    case VK_QUERY_PIPELINE_STATISTIC_CLIPPING_INVOCATIONS_BIT:
+        return "Clipping stage primitives processed";
+    case VK_QUERY_PIPELINE_STATISTIC_CLIPPING_PRIMITIVES_BIT:
+        return "Clipping stage primitives output";
+    case VK_QUERY_PIPELINE_STATISTIC_FRAGMENT_SHADER_INVOCATIONS_BIT:
+        return "Fragment shader invocations";
+    case VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_INVOCATIONS_BIT:
+        return "Geometry shader invocations";
+    case VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_PRIMITIVES_BIT:
+        return "Geometry assembly primitives count";
+    case VK_QUERY_PIPELINE_STATISTIC_TESSELLATION_CONTROL_SHADER_PATCHES_BIT:
+        return "Tessellation control shader patch invocations";
+    case VK_QUERY_PIPELINE_STATISTIC_TESSELLATION_EVALUATION_SHADER_INVOCATIONS_BIT:
+        return "Tessellation evaluation shader invocations";
+    case VK_QUERY_PIPELINE_STATISTIC_COMPUTE_SHADER_INVOCATIONS_BIT:
+        return "Compute shader invocations";
+    default:
+        break;
+    }
+    return "Unknown";
+}
+
+QueryPool::QueryPool(const Device &device, const std::string &name)
+    : QueryPool(device, name, default_pipeline_stats_flag_bits){};
+
+QueryPool::QueryPool(const Device &device, const std::string &name,
+                     const std::vector<VkQueryPipelineStatisticFlagBits> &pipeline_stats_flag_bits)
+    : m_device(device) {
+    assert(m_device.device());
+    assert(!name.empty());
+
+    // We must first check if pipeline query statistics are available.
+    vkGetPhysicalDeviceFeatures(m_device.physical_device(), &m_device_features);
+
+    if (!static_cast<bool>(m_device_features.pipelineStatisticsQuery)) {
```

Comment:
```suggestion
    if (m_device_features.pipelineStatisticsQuery == VK_FALSE) {
```
`static_cast<bool>` is almost never the best thing to do, even though it's the first thing clang-tidy suggests. Same for the ones in the `validate_pipeline_stats_flag_bits` function too.

Reply:
Done.